addons/product/models/product_template.py

```diff
@@ -1125,9 +1125,10 @@ def _get_possible_combinations(self, parent_combination=None, necessary_values=N
             yield necessary_values
 
         product_template_attribute_values_per_line = [
-            ptal.product_template_value_ids
+            ptal.product_template_value_ids._only_active()
             for ptal in attribute_lines
         ]
+        product_template_attribute_values_per_line = list(filter(None, product_template_attribute_values_per_line))
```

Comment:
Why this line?
It feels like it should make no sense to have an attribute line without values in the first place, if that is indeed what it is filtering?

And if it is actually needed, I guess `if ptal.product_template_value_ids` should just be added in the list generation just above instead.

Reply:
Couldn't we have a situation when there is an attribute with all values archived? I'm not sure that it's possible, but  it seems better to do the filtering in stable (e.g. in case of some customization or even misusing of the feature).

Reply:
Possible, I guess?

But that basically means the template is not usable because `_is_combination_possible_by_config` should always return `False` in that case, so if that happens you might as well directly `return` here if you are trying to optimize.

Reply:
Maybe `_cartesian_product` returns no result if you give it an empty list? I would hope that is the case, and then you don't need to worry about it.

Reply:
`_cartesian_product` looks complicated, so I prefer to don't change it and filter data outside. 
it seems it doesn't support empty recordset because of the following line:
```py
            current_line_values = product_template_attribute_values_per_line[line_index]
```

Reply:
`line_index` is the line number so that is always valid, it will just mean `current_line_values` will be empty recordset.

What might be broken though is the following line `current_ptav = current_line_values[current_ptav_index]` when `current_ptav_index` is -1 and the `current_line_values` is empty.

I think the proper guard is to iterate values of `product_template_attribute_values_per_line` at the top of `_cartesian_product` and return immediately if one of the line is empty.

Please don't guess, just add a test to cover this case and make the suggested fix if the issue happens.

Reply:
I've checked and it's impossible to have a line with empty active lines

![screenshot-devel 14 localhost_14069-2021 07 23-15_57_53](https://user-images.githubusercontent.com/186131/126792636-b7b2821f-a0df-47e8-9101-ec4e26f21590.png)


I've deleted to line of code

Reply:
Thanks for checking, indeed it's not possible by normal config.

Unfortunately the constraint takes into account all values, even archived one (see `_check_valid_values` on line model). So it's still possible to mess it up by having a line with a value and then archiving the value afterwards somehow.

It depends how careful we want to be with the fix, but I think your original feeling is legitimate and it's better to still add the guard and a test to cover its need.

Reply:
> Unfortunately the constraint takes into account all values, even archived one (see `_check_valid_values` on line model)

It is either incorrect or you mean something else.

Values `product.attribute.value` cannot be archived [1] (there is no `active` field). The constrain `_check_valid_values` [2] should be enough to prevent empty values in `ptal.product_template_value_ids._only_active()`, because `_only_active` filters by field `ptav_active` [3] and `ptav_active`  can be set to False either via customization (i.e. bad customization, that leads to inconsistent data) or by removing  `product.attribute.value` [4] [5].

Anyway, I don't quite get you point. If I understand you correctly, you ask me write a test that will check the code against inconsistent data and then pass that test by changing Cartesian products method. Why just don't keep original line of code that explicitly deletes inconsistent data and hence doesn't require a test?

[1]

https://github.com/odoo/odoo/blob/f9b12a59ea618fa598f0ffb6869f027c6a39bc2d/addons/product/models/product_attribute.py#L85-L109

[2]

https://github.com/odoo/odoo/blob/f9b12a59ea618fa598f0ffb6869f027c6a39bc2d/addons/product/models/product_attribute.py#L181-L187

[3]

https://github.com/odoo/odoo/blob/f9b12a59ea618fa598f0ffb6869f027c6a39bc2d/addons/product/models/product_attribute.py#L505-L506

[4]

https://github.com/odoo/odoo/blob/f9b12a59ea618fa598f0ffb6869f027c6a39bc2d/addons/product/models/product_attribute.py#L495

[5]

https://github.com/odoo/odoo/blob/d6cf58ae6fbb32c2d1dae1f308e430d4aa4488d5/addons/product/models/product_attribute.py#L361